src/components/tree-view/tree-view.tsx

```diff
@@ -25,37 +28,77 @@ export interface TreeViewProps {
     selectedItem?: TreeItemData;
 }
 
+export interface TreeItemState {
+    isSelected: boolean;
+}
+
+export type StateMap = Map<TreeItemData, TreeItemState>;
+
 const itemIdPrefix = 'TREE_ITEM';
 
-export function TreeItem({ item, itemRenderer, onItemClick, isSelected }: TreeItemProps): JSX.Element {
+export function TreeItem({ item, itemRenderer, onItemClick, stateMap, state }: TreeItemProps): JSX.Element {
```

Comment:
for stateless functional comps I myself used ```const TreeItem: React.SFC<TreeItemProps> = (props) => {}``` . I'm not too opinionated about it but let's decide on what seems more expressive?